File: backend/app/schemas/resume.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Any, Union
from datetime import datetime
from enum import Enum

class ResumeStatus(str, Enum):
    """Resume processing status"""
    UPLOADED = "uploaded"
    PARSING = "parsing"
    PARSED = "parsed"
    SCORING = "scoring"
    SCORED = "scored"
    ERROR = "error"

class ResumeFileType(str, Enum):
    """Supported resume file types"""
    PDF = "application/pdf"
    DOCX = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
    DOC = "application/msword"
    TEX = "application/x-tex"
    TEX_ALT = "text/x-tex"
    PLAIN = "text/plain"  # For .tex files that might come as text/plain

class UploadUrlRequest(BaseModel):
    """Request for presigned upload URL"""
    filename: str = Field(..., min_length=1, max_length=255)
    content_type: str = Field(..., description="MIME type of the file")
    file_size: int = Field(..., gt=0, le=10_485_760, description="File size in bytes (max 10MB)")

class UploadUrlResponse(BaseModel):
    """Response with presigned upload URL"""
    upload_url: str
    resume_id: str
    storage_path: str
    expires_in: int = 3600  # seconds

class UploadCallbackRequest(BaseModel):
    """Notification that upload is complete"""
    resume_id: str
    storage_path: str

# New models for wizard resume creation
class ContactInfo(BaseModel):
    """Contact information"""
    name: str
    email: str
    phone: Optional[str] = None
    location: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None
    leetcode: Optional[str] = None
    codechef: Optional[str] = None
    hackerrank: Optional[str] = None
    website: Optional[str] = None

class ExperienceEntry(BaseModel):
    """Work experience entry"""
    id: Optional[str] = None
    company: str
    position: str
    title: Optional[str] = None
    location: Optional[str] = None
    startDate: str
    endDate: Optional[str] = None
    description: Optional[str] = None
    current: Optional[bool] = False

class EducationEntry(BaseModel):
    """Education entry"""
    id: Optional[str] = None
    school: str
    degree: str
    field: Optional[str] = None
    location: Optional[str] = None
    startDate: Optional[str] = None
    endDate: Optional[str] = None
    gpa: Optional[str] = None
    gradeType: Optional[str] = "GPA"  # GPA, CGPA, or Percentage
    description: Optional[str] = None

class SkillCategory(BaseModel):
    """Skill category with items"""
    category: str
    items: List[str] = []

class SkillsData(BaseModel):
    """Skills data - supports both old {technical, soft} and new [{category, items}] format"""
    # Old format fields (optional)
    technical: Optional[List[str]] = None
    soft: Optional[List[str]] = None
    # New format: list of categories 
    categories: Optional[List[SkillCategory]] = None
    
    @classmethod
    def from_list(cls, skills_list: List[dict]) -> 'SkillsData':
        """Convert list format to SkillsData"""
        categories = [SkillCategory(**s) for s in skills_list if 'category' in s]
        return cls(categories=categories)
    
    def to_list(self) -> List[dict]:
        """Convert to list format for storage/template"""
        result = []
        if self.categories:
            result.extend([{'category': c.category, 'items': c.items} for c in self.categories])
        if self.technical:
            result.append({'category': 'Technical', 'items': self.technical})
        if self.soft:
            result.append({'category': 'Soft Skills', 'items': self.soft})
        return result

class ProjectEntry(BaseModel):
    """Project entry"""
    id: Optional[str] = None
    name: str
    description: Optional[str] = None
    technologies: Optional[str] = None
    url: Optional[str] = None
    startDate: Optional[str] = None
    endDate: Optional[str] = None

class CertificationEntry(BaseModel):
    """Certification entry"""
    id: Optional[str] = None
    name: str
    issuer: str
    date: Optional[str] = None
    credentialId: Optional[str] = None
    url: Optional[str] = None

class LanguageEntry(BaseModel):
    """Language entry"""
    id: Optional[str] = None
    name: str
    proficiency: str

class AchievementEntry(BaseModel):
    """Achievement entry"""
    id: Optional[str] = None
    title: str
    description: Optional[str] = None
    date: Optional[str] = None

class CreateResumeRequest(BaseModel):
    """Request to create resume from wizard data"""
    template: Optional[str] = "resume_1"  # Template ID: resume_1, resume_2, etc.
    contact: ContactInfo
    summary: str
    experience: List[ExperienceEntry] = []
    education: List[EducationEntry] = []
    skills: Union[SkillsData, List[SkillCategory], List[dict]] = []
    projects: List[ProjectEntry] = []
    certifications: List[CertificationEntry] = []
    languages: List[LanguageEntry] = []
    achievements: List[AchievementEntry] = []
    
    @field_validator('skills', mode='before')
    @classmethod
    def normalize_skills(cls, v):
        """Accept both old format and new list format for skills"""
        if v is None:
            return []
        if isinstance(v, list):
            # New format: list of {category, items}
            return v
        if isinstance(v, dict):
            # Old format: {technical: [], soft: []}
            # Convert to new format
            result = []
            if v.get('technical'):
                result.append({'category': 'Technical', 'items': v['technical']})
            if v.get('soft'):
                result.append({'category': 'Soft Skills', 'items': v['soft']})
            if v.get('categories'):
                result.extend(v['categories'])
            return result
        return v

class ResumeMetadata(BaseModel):
    """Resume metadata stored in Firestore"""
    resume_id: str
    owner_uid: str
    filename: str
    original_filename: str
    content_type: str
    file_size: int
    storage_path: str
    storage_url: Optional[str] = None
    status: ResumeStatus = ResumeStatus.UPLOADED
    created_at: datetime
    updated_at: datetime
    template: Optional[str] = "resume_1"  # Template ID used for this resume
    
    # Parsing results (populated after parsing)
    parsed_text: Optional[str] = None
    contact_info: Optional[dict] = None
    skills: Optional[Any] = None  # Can be list [{category, items}] or dict {technical, soft}
    sections: Optional[Any] = None  # Dynamic sections - can be list or dict for backwards compatibility
    experience: Optional[List[dict]] = None
    projects: Optional[List[dict]] = None
    education: Optional[List[dict]] = None
    certifications: Optional[List[dict]] = None
    languages: Optional[List[dict]] = None
    achievements: Optional[List[dict]] = None
    layout_type: Optional[str] = None
    parsed_at: Optional[datetime] = None
    
    # Scoring results (populated after scoring)
    latest_score: Optional[float] = None
    latest_score_id: Optional[str] = None
    
    # Error tracking
    error_message: Optional[str] = None
    
    @field_validator('sections', mode='before')
    @classmethod
    def convert_sections(cls, v):
        """Convert dict to list or keep as list for backwards compatibility"""
        if v is None:
            return []
        if isinstance(v, dict):
            # Old format was dict, convert to empty list
            return []
        if isinstance(v, list):
            return v
        return []

class ResumeListItem(BaseModel):
    """Resume item in list view"""
    resume_id: str
    filename: str
    original_filename: str
    file_size: int
    status: ResumeStatus
    created_at: datetime
    latest_score: Optional[float] = None
    
class ResumeListResponse(BaseModel):
    """List of user's resumes"""
    resumes: List[ResumeListItem]
    total: int

class ResumeDetailResponse(BaseModel):
    """Detailed resume information"""
    resume_id: str
    filename: str
    original_filename: str
    content_type: str
    file_size: int
    storage_url: Optional[str] = None
    status: ResumeStatus
    created_at: datetime
    updated_at: datetime
    parsed_text: Optional[str] = None
    contact_info: Optional[dict] = None
    skills: Optional[Any] = None  # Can be list [{category, items}] or dict {technical, soft}
    sections: Optional[Any] = None  # Dynamic sections - can be list or dict for backwards compatibility
    experience: Optional[List[dict]] = None
    projects: Optional[List[dict]] = None
    education: Optional[List[dict]] = None
    certifications: Optional[List[dict]] = None
    languages: Optional[List[dict]] = None
    achievements: Optional[List[dict]] = None
    layout_type: Optional[str] = None
    parsed_at: Optional[datetime] = None
    latest_score: Optional[float] = None
    template: Optional[str] = "resume_1"
    error_message: Optional[str] = None
    
    @field_validator('sections', mode='before')
    @classmethod
    def convert_sections(cls, v):
        """Convert dict to list or keep as list for backwards compatibility"""
        if v is None:
            return []
        if isinstance(v, dict):
            # Old format was dict, convert to empty list
            return []
        if isinstance(v, list):
            return v
        return []

class CreateResumeVersionRequest(BaseModel):
    """Request to create a new version of a resume"""
    resume_id: str
    job_role: Optional[str] = None
    company: Optional[str] = None
    resume_json: dict

class TailorResumeRequest(BaseModel):
    """Request to tailor a resume for a job description"""
    job_description: str
    job_role: Optional[str] = None
    company: Optional[str] = None

class ResumeVersionResponse(BaseModel):
    """Response for created version"""
    version_id: str
    version_name: str
    created_at: datetime

class ResumeVersionDetailResponse(ResumeVersionResponse):
    """Detailed response for a version including the full JSON"""
    resume_json: dict

File: frontend/src/types/resume.ts
/**
 * Resume data types for the editor
 */

export interface ContactInfo {
  fullName: string;
  email: string;
  phone: string;
  location: string;
  linkedin?: string;
  github?: string;
  portfolio?: string;
  otherLinks?: string[];
}

export interface Experience {
  id: string;
  company: string;
  position: string;
  title: string;  // Alias for position
  location: string;
  startDate: string;
  endDate: string;
  current: boolean;
  description: string;
  highlights: string[];
}

export interface Education {
  id: string;
  institution: string;
  degree: string;
  field: string;
  location: string;
  startDate: string;
  endDate: string;
  gpa?: string;
  honors?: string;
}

export interface Project {
  id: string;
  name: string;
  description: string;
  technologies: string[];
  link?: string;
  highlights: string[];
  startDate?: string;
  endDate?: string;
}

export interface Skill {
  category: string;
  items: string[];
}

export interface Certification {
  id: string;
  name: string;
  issuer: string;
  date: string;
  credentialId?: string;
  url?: string;
}

export interface Achievement {
  id: string;
  title: string;
  text?: string;
  description: string;
  result?: string;
  date: string;
}

export interface Hackathon {
  id: string;
  name: string;
  achievement: string;
  date: string;
  description?: string;
}

export interface Workshop {
  id: string;
  name: string;
  role: string;
  description: string;
  date?: string;
}

export interface Publication {
  id: string;
  title: string;
  publisher: string;
  date: string;
  link?: string;
  authors?: string;
}

export interface Volunteer {
  id: string;
  organization: string;
  role: string;
  date: string;
  description: string;
}

export interface Language {
  id: string;
  language: string;
  proficiency: string;
}

// Dynamic section type for flexible resume structure
export type SectionType =
  | 'contact'
  | 'summary'
  | 'experience'
  | 'education'
  | 'skills'
  | 'projects'
  | 'certifications'
  | 'achievements'
  | 'hackathons'
  | 'workshops'
  | 'publications'
  | 'volunteer'
  | 'languages'
  | 'custom';

// Human-readable labels for each section type
export const SECTION_LABELS: Record<SectionType, string> = {
  contact: 'Contact Information',
  summary: 'Professional Summary',
  experience: 'Work Experience',
  education: 'Education',
  skills: 'Skills',
  projects: 'Projects',
  certifications: 'Certifications',
  achievements: 'Achievements',
  hackathons: 'Hackathons & Competitions',
  workshops: 'Workshops',
  publications: 'Publications',
  volunteer: 'Volunteer Work',
  languages: 'Languages',
  custom: 'Custom Section',
};

// Available sections that can be added dynamically
export const AVAILABLE_SECTIONS: SectionType[] = [
  'experience',
  'education',
  'skills',
  'projects',
  'certifications',
  'achievements',
  'hackathons',
  'workshops',
  'publications',
  'volunteer',
  'languages',
  'custom',
];

export interface DynamicSection {
  id: string;
  type: SectionType;
  title: string;
  items: Record<string, unknown>[];
  order: number;
}

export interface ResumeData {
  id?: string;
  userId?: string;
  template?: string;  // Template ID (resume_1 through resume_7)
  contact: ContactInfo;
  summary: string;
  experience: Experience[];
  education: Education[];
  skills: Skill[];
  projects: Project[];
  certifications?: Certification[];
  achievements?: Achievement[];
  hackathons?: Hackathon[];
  workshops?: Workshop[];
  publications?: Publication[];
  volunteer?: Volunteer[];
  languages?: Language[];
  // Dynamic sections from Gemini parsing
  sections?: DynamicSection[];
  customSections?: { [key: string]: any[] };
  createdAt?: Date;
  updatedAt?: Date;
  originalFileUrl?: string; // URL to the originally uploaded file
}

export const createEmptyResume = (): ResumeData => ({
  contact: {
    fullName: '',
    email: '',
    phone: '',
    location: '',
  },
  summary: '',
  experience: [],
  education: [],
  skills: [],
  projects: [],
  certifications: [],
  achievements: [],
  hackathons: [],
  workshops: [],
  publications: [],
  volunteer: [],
  languages: [],
  sections: [],
  customSections: {},
});

export const createEmptyExperience = (): Experience => ({
  id: crypto.randomUUID(),
  company: '',
  position: '',
  title: '',
  location: '',
  startDate: '',
  endDate: '',
  current: false,
  description: '',
  highlights: [],
});

export const createEmptyEducation = (): Education => ({
  id: crypto.randomUUID(),
  institution: '',
  degree: '',
  field: '',
  location: '',
  startDate: '',
  endDate: '',
});

export const createEmptyProject = (): Project => ({
  id: crypto.randomUUID(),
  name: '',
  description: '',
  technologies: [],
  highlights: [],
});

export const createEmptySkill = (): Skill => ({
  category: '',
  items: [],
});

export const createEmptyCertification = (): Certification => ({
  id: crypto.randomUUID(),
  name: '',
  issuer: '',
  date: '',
});

export const createEmptyAchievement = (): Achievement => ({
  id: crypto.randomUUID(),
  title: '',
  description: '',
  date: '',
});

export const createEmptyHackathon = (): Hackathon => ({
  id: crypto.randomUUID(),
  name: '',
  achievement: '',
  date: '',
  description: '',
});

export const createEmptyWorkshop = (): Workshop => ({
  id: crypto.randomUUID(),
  name: '',
  role: '',
  description: '',
  date: '',
});

export const createEmptyPublication = (): Publication => ({
  id: crypto.randomUUID(),
  title: '',
  publisher: '',
  date: '',
  link: '',
});

export const createEmptyVolunteer = (): Volunteer => ({
  id: crypto.randomUUID(),
  organization: '',
  role: '',
  date: '',
  description: '',
});

export const createEmptyLanguage = (): Language => ({
  id: crypto.randomUUID(),
  language: '',
  proficiency: '',
});
